sonarts-core/src/runner/sonartsServer.ts

```diff
@@ -0,0 +1,103 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+import * as net from "net";
+import * as path from "path";
+import * as os from "os";
+import * as fs from "fs";
+import * as ts from "typescript";
+import { getIssues } from "./rules";
+import { FileCache, createService } from "./languageService";
+import { parseTsConfig } from "../utils/parser";
+
+export function start(port: number) {
+  const EMPTY_ANSWER = '{"issues":[]}';
+  const fileCache = new FileCache();
+  // key is ts file path, value is corresponding tsconfig path
+  const tsConfigCache: Map<string, string> = new Map();
+  const servicesPerTsconfig: Map<string, ts.LanguageService> = new Map();
+
+  const server = net.createServer(socket => {
+    socket.on("data", data => {
+      const request = JSON.parse(data.toString());
+
+      if (request.operation === "analyze") {
+        const { file, rules, content } = request;
+
+        fileCache.newContent({ file, content });
+
+        let tsConfig;
+        if (tsConfigCache.has(file)) {
+          tsConfig = tsConfigCache.get(file)!;
+        } else {
+          tsConfig = getTsConfig(file);
+          if (tsConfig) {
+            tsConfigCache.set(file, tsConfig);
+          } else {
+            console.error("No tsconfig.json file found for " + file);
+            socket.write(EMPTY_ANSWER);
+            return;
+          }
+        }
+
+        let service = servicesPerTsconfig.get(tsConfig);
+        if (!service) {
+          const { files, options } = parseTsConfig(tsConfig);
+          service = createService(files, options, fileCache);
+          servicesPerTsconfig.set(tsConfig, service);
+        }
+
+        const program = service.getProgram();
+        const sourceFile = program.getSourceFile(file);
+        if (!sourceFile) {
+          console.error(`No SourceFile found for file ${file} with configuration ${tsConfig}`);
+          socket.write(EMPTY_ANSWER);
+          return;
+        }
+        const issues = getIssues(rules, program, sourceFile);
+        socket.write(JSON.stringify(issues));
+      } else {
+        console.error("Unknown operation for SonarTS Server: " + request.operation);
+        socket.write(EMPTY_ANSWER);
+      }
+    });
+  });
+  server.listen(port, "localhost");
+  console.log("SonarTS Server started on port " + port + " from folder " + __dirname);
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Use a template literal instead of concatenating with a string literal. ([typescript:S3512](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS3512))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=sonarts&branch=issue-351&issues=AWD5U5dREFfjxkQcPdca&open=AWD5U5dREFfjxkQcPdca)

[ISSUE_KEY]: AWD5U5dREFfjxkQcPdca

Reply:
This issue has been resolved by a user in SonarQube

[ISSUE_STATUS]: RESOLVED